src/lnp/session/noise/conduit.rs

```diff
@@ -0,0 +1,630 @@
+            chacha::decrypt(
+                &self.receiving_key,
+                self.receiving_nonce as u64,
+                &[0; 0],
+                encrypted_length,
+                &mut length_bytes,
+            )?;
+
+            self.increment_nonce();
+
+            // the message length
+            //byte_utils::slice_to_be16(&length_bytes) as usize
+            u16::from_be_bytes(length_bytes) as usize
+        };
+
+        let message_end_index = TAGGED_MESSAGE_LENGTH_HEADER_SIZE
+            + message_length
+            + chacha::TAG_SIZE;
+
+        if buffer.len() < message_end_index {
+            self.pending_message_length = Some(message_length);
+            return Ok((None, 0));
+        }
+
+        self.pending_message_length = None;
+
+        let encrypted_message =
+            &buffer[TAGGED_MESSAGE_LENGTH_HEADER_SIZE..message_end_index];
+        let mut message = vec![0u8; message_length];
+
+        chacha::decrypt(
+            &self.receiving_key,
+            self.receiving_nonce as u64,
+            &[0; 0],
+            encrypted_message,
+            &mut message,
+        )?;
+
+        self.increment_nonce();
+
+        Ok((Some(message), message_end_index))
+    }
+
+    fn increment_nonce(&mut self) {
+        Conduit::increment_nonce(
+            &mut self.receiving_nonce,
+            &mut self.receiving_chaining_key,
+            &mut self.receiving_key,
+        );
+    }
+
+    // Used in tests to determine whether or not excess bytes entered the
+    // conduit without needing to bring up infrastructure to properly encode
+    // it
+    #[cfg(test)]
+    pub fn read_buffer_length(&self) -> usize {
+        match &self.read_buffer {
+            &Some(ref vec) => vec.len(),
+            &None => 0,
+        }
+    }
+}
+
+impl Iterator for Decryptor {
+    type Item = Result<Option<Vec<u8>>, Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.poisoned {
+            return None;
+        }
+
+        match self.decrypt_single_message(None) {
+            Ok(Some(result)) => Some(Ok(Some(result))),
+            Ok(None) => None,
+            Err(e) => {
+                self.poisoned = true;
+                Some(Err(e))
+            }
+        }
+    }
+}
+
+impl Decrypt for Decryptor {
+    type Error = HandshakeError;
+    fn decrypt(
+        &mut self,
+        buffer: impl Borrow<[u8]>,
+    ) -> Result<Vec<u8>, Self::Error> {
+        match self.decrypt_single_message(Some(buffer.borrow())) {
+            Ok(Some(data)) => Ok(data),
+            Ok(None) => Ok(Vec::new()),
+            Err(e) => Err(HandshakeError::ChaCha20(e)),
+        }
+    }
+}
+
+/// Returned after a successful handshake to encrypt and decrypt communication
+/// with peer nodes. It should not normally be manually instantiated.
+/// Automatically handles key rotation.
+/// For decryption, it is recommended to call `decrypt_message_stream` for
+/// automatic buffering.
+pub struct Conduit {
```

Comment:
Many be name it as `NoiseEncryption` or `LnEncryption`? To match `NoEncryption` name?

Reply:
Noted. Going for `NoiseEncryption`.